### PR TITLE
CI: skip azp and circleCI logic

### PR DESCRIPTION
**Pull request by tupui**
I propose to add a `[skip azp]` and `[skip circle]` to respectively skip Azure and CircleCI.
Code taken from SciPy.

Other CIs can already be skipped with: `[skip actions] [skip travis]`

This would allow finer control on which part of the CI is being run.

### CODE REVIEW

**Comment by charris** [comment]
I don't see anything wrong with this, but skipping is seldom used in practice. And sometimes skipping documentation tests misses bugs.

**Comment by seberg** [comment]
Not sure the circle CI thing worked...  I am fishing for bread crumbs now, but the probability of me not forgetting the right skips might dramatically increase if we had only tags like `[CI docs only]` or `[CI no docs]` (could be labels as well, might remember to put them or at least add them when you iterate a lot).
Nobody needs to pick out exactly travis, etc...

**Comment by tupui** [comment]
> I don't see anything wrong with this, but skipping is seldom used in practice. And sometimes skipping documentation tests misses bugs.

I know, same "issue" in SciPy. To me it should not be the case. We should care more about what we consume. That's a larger discussion for sure. Seeing things like PyPi stats continuously increasing is not sustainable (arguably driven by CI all over the place). At one point it will just explode, Cloudflare will not pay an infinite bill/use of their infra.

> Not sure the circle CI thing worked...

Ah yes, this is due to the different merge strategy that SciPy has.

> I am fishing for bread crumbs now, but the probability of me not forgetting the right skips might dramatically increase if we had only tags like [CI docs only] or [CI no docs] (could be labels as well, might remember to put them or at least add them when you iterate a lot).
Nobody needs to pick out exactly travis, etc...

I am 100% there. IMHO CI should not run automatically for PR and be conditioned to a maintainer dropping a [combination of] label.

**Comment by charris** [comment]
> IMHO CI should not run automatically for PR

One of the first things I look for in a PR are passing tests. For documentation only I would expect lint and circleCI to run at the least. Note that we have two lint checks, we could probably do away with one of them. What I would really like is checks that depended on which files were touched.

**Comment by seberg** [comment]
If PRs labelled (or auto-labeled, i.e. starting with `DOC:` and not also any of the other categories) only run the tests:
- lint
- smoke test (maybe just to run some real test)
- CircleCI
- Possibly the refguide check test runs (which probably make the smoke test irrelevant)

that would be pretty cool.  The opposite direction could skip CircleCI (not the refcheck ideally).

We could also check more really, then we would just rely on the occasionally odd thing to be found only after the fact, which is not super terrible.

But... is it easy to use labels?  And if the labels change, re-triggering CI might be annoying?

**Comment by seberg** [comment]
Are you happy with this? Then we should probably just put it in for now...

> put a label to ask the CI to run and the label is automatically removed by the CI when it does run

The technique may be interesting, although I would not like it if maintainers need to constantly followup on someone iterating a PR.  It may be possible to convince me that the complete test run should label only (but even then, maybe no auto-remove?).

**Comment by tupui** [comment]
> Are you happy with this? Then we should probably just put it in for now...

I am, but you folks use NumPy's CI way more than I occasionally do 😅 So as you like it.

> > put a label to ask the CI to run and the label is automatically removed by the CI when it does run
> 
> The technique may be interesting, although I would not like it if maintainers need to constantly followup on someone iterating a PR. It may be possible to convince me that the complete test run should label only (but even then, maybe no auto-remove?).

Yeah it comes down to a workflow decision and how you want to work/interact with everyone.

**Comment by rgommers** [comment]
+1 for this PR adding `[skip azp]` and `[skip circle]`, quite useful.

Re running less CI by default, I agree with comments of @charris and @seberg. A smaller default set and a single label-based extra like `ci-all` would perhaps work. There's a lot of design flexibility though, and it's a pain to deal with the infra (needs a bot + need to support multiple CI providers), so I'd be happy to not have to think about that.

File: .circleci/config.yml [hunk 1]
```diff
@@ -18,6 +18,17 @@ jobs:
...
     steps:
       - checkout:
+
+      - run:
+          name: Check-skip
+          command: |
+            export git_log=$(git log --max-count=1 --pretty=format:"%B" | tr "\n" " ")
+            echo "Got commit message:"
+            echo "${git_log}"
+            if [[ -v CIRCLE_PULL_REQUEST ]] && ([[ "$git_log" == *"[skip circle]"* ]] || [[ "$git_log" == *"[circle skip]"* ]]); then
```
Comment (charris) [inline .circleci/config.yml:28]: Could you break this long line using line continuation?

File: .circleci/config.yml [hunk 2]
```diff
@@ -18,6 +18,17 @@ jobs:
...
+              echo "Skip detected, exiting job ${CIRCLE_JOB} for PR ${CIRCLE_PULL_REQUEST}."
+              circleci-agent step halt;
+            fi
```
Comment (tupui) [inline .circleci/config.yml:31]:
```suggestion
            if [[ -v CIRCLE_PULL_REQUEST ]] && \
               ([[ "$git_log" == *"[skip circle]"* ]] || \
                [[ "$git_log" == *"[circle skip]"* ]])
            then
              echo "Skip detected, exiting job ${CIRCLE_JOB} for PR ${CIRCLE_PULL_REQUEST}."
              circleci-agent step halt;
            fi
```

File: .circleci/config.yml (hunk 1 above)
Comment (tupui) [reply .circleci/config.yml]: Done

**Comment by tupui** [comment]
@charris any news there?

**Comment by mattip** [comment]
The list of `[skip ...]` commands is [here](https://numpy.org/devdocs/dev/development_workflow.html?highlight=ci+skip#commands-to-skip-continuous-integration), the source is [here](https://github.com/numpy/numpy/blob/b6a3e837785eac58a2f68e126f4db7895ca047b3/doc/source/dev/development_workflow.rst?plain=1#L191). I think we already have `[skip azurepipelines]`, how does the current PR interact with that?

**Comment by tupui** [comment]
@mattip I can update this listing. The `[skip azurepipelines]` is provided by Azure and only works when on `main`... They have tons of reports of user asking to make this work in PRs, but no. So this adds the ability to skip on PRs as well (like any other CI provider). If this is merged, I would just advise to only use `[skip azp]`

**Comment by mattip** [comment]
> I would just advise to only use `[skip azp]`

+1 for adjusting the documentation appropriately. What [scipy has](https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping) seems good. A future PR could create a page like that for NumPy.

**Comment by tupui** [comment]
Done 😃

**Comment by mattip** [comment]
Thanks @tupui

**Pull request merged**